### PR TITLE
charm dispatch changes in uniter

### PR DESCRIPTION
**Pull request by hmlanigan**
### Checklist

 - [x] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?
 - [ ] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?
 - [ ] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

As part of uniter++ work, add JUJU_DISPATH_PATH env var for charms to use.  If a 'dispatch' script is located at the top level of the charm, call it.  Otherwise normal behavior of calling hooks/<name> or action/<name>.  The debug-hook welcome message has been updated accordingly.  New charms for both IAAS and CAAS models have been included to test this functionality.

This change does not cover actions on k8s charms.  That functionality will be implemented in a new PR shortly.

## QA steps

For IAAS:
```console
$ juju bootstrap localhost test

# The ubuntu charm should continue to operate as normal.
$ juju deploy ubuntu

$ juju deploy ./testcharms/charm-repo/bionic/ubuntu-plus/

# After the unit is up and running and first update-status called:
$ juju debug-log --replay | grep juju-log | grep Calling
unit-ubuntu-plus-0: 20:57:34 INFO unit.dispatch/0.juju-log Calling hooks/install, from dispatch
unit-ubuntu-plus-0: 21:02:44 INFO unit.dispatch/0.juju-log Calling hooks/update-status, from dispatch

# Run action to move dispatch file, view diff in juju-log calls:
$ juju run-action ubuntu-plus/0 no-dispatch filename=testfile
Action queued with id: "2"
$ juju show-action-output 2
...
  outcome: success
  result-map:
    message: dispatch is now called testfile
    time-completed: Thu Feb 27 23:34:50 UTC 2020
status: completed
...

# Wait for next run of update-status hook then:
$ juju debug-log | grep juju-log | grep Calling
# stop seeing "from dispatch"
$ juju debug-log --replay --include unit-ubuntu-plus-0 | grep juju-log | grep update-status | tail
unit-ubuntu-plus-0: 23:31:14 INFO unit.ubuntu-plus/0.juju-log Calling hooks/update-status, from dispatch
unit-ubuntu-plus-0: 23:31:15 INFO unit.ubuntu-plus/0.juju-log Calling actions/no-dispatch, from dispatch
unit-ubuntu-plus-0: 23:36:26 INFO unit.ubuntu-plus/0.juju-log Hello from update-status.

# verify with:
$ juju debug-log  --include unit-ubuntu-plus-0 --include-module  juju.worker.uniter.operation
unit-ubuntu-plus-0: 17:48:15 INFO juju.worker.uniter.operation ran "install" hook (via hook dispatching script: dispatch)
unit-ubuntu-plus-0: 17:48:15 INFO juju.worker.uniter.operation ran "leader-elected" hook (via hook dispatching script: dispatch)
unit-ubuntu-plus-0: 17:48:15 INFO juju.worker.uniter.operation ran "config-changed" hook (via hook dispatching script: dispatch)
unit-ubuntu-plus-0: 17:48:15 INFO juju.worker.uniter.operation ran "start" hook (via hook dispatching script: dispatch)
unit-ubuntu-plus-0: 17:53:14 INFO juju.worker.uniter.operation ran "update-status" hook (via hook dispatching script: dispatch)
unit-ubuntu-plus-0: 17:57:47 INFO juju.worker.uniter.operation ran "update-status" hook (via explicit, bespoke hook script)
unit-ubuntu-plus-0: 18:02:50 INFO juju.worker.uniter.operation ran "update-status" hook (via explicit, bespoke hook script)
```

For CAAS (assuming that microk8s is install and configured on the test machine):
```console
$ make microk8s-operator-update
$ juju bootstrap microk8s mk8s --no-gui
$ juju add-model testme

$ docker pull ubuntu
$ cd testcharms/charm-repo/kubernetes/
$ mkdir ~/charms
$ charm build -o ~/charms ubuntu-plus
$ ( cd ~ ; juju deploy ~/charms/builds/ubuntu-plus-k8s --resource ubuntu_image=ubuntu )

# After the unit is up and running and first update-status called:
$ juju debug-log --replay  | grep Call
application-ubuntu-plus-k8s: 18:33:20 INFO unit.ubuntu-plus-k8s/0.juju-log Calling hooks/install, from dispatch
application-ubuntu-plus-k8s: 18:33:23 INFO unit.ubuntu-plus-k8s/0.juju-log Calling hooks/leader-elected, from dispatch
application-ubuntu-plus-k8s: 18:33:24 INFO unit.ubuntu-plus-k8s/0.juju-log Calling hooks/config-changed, from dispatch
application-ubuntu-plus-k8s: 18:33:25 INFO unit.ubuntu-plus-k8s/0.juju-log Calling hooks/start, from dispatch
application-ubuntu-plus-k8s: 18:38:19 INFO unit.ubuntu-plus-k8s/0.juju-log Calling hooks/update-status, from dispatch
```

## Documentation changes

As part of 2.8 release.

### CODE REVIEW

File: worker/uniter/runner/debug/server_test.go [hunk 1]
```diff
@@ -238,6 +241,5 @@ func (s *DebugHooksServerSuite) verifyEnvshFile(c *gc.C, envshPath string, hookN
 	c.Assert(err, jc.ErrorIsNil)
 	contents := string(data)
 	c.Assert(contents, jc.Contains, fmt.Sprintf("JUJU_UNIT_NAME=%q", s.ctx.Unit))
-	c.Assert(contents, jc.Contains, fmt.Sprintf("JUJU_HOOK_NAME=%q", hookName))
 	c.Assert(contents, jc.Contains, fmt.Sprintf(`PS1="%s:%s %% "`, s.ctx.Unit, hookName))
```
Comment (hmlanigan) [inline worker/uniter/runner/debug/server_test.go]: The test passes for this.  But when I do a debug-hook and get into the context, this isn't the the prompt.

File: worker/uniter/runner/runner.go [hunk 2]
```diff
@@ -549,10 +596,27 @@ func (runner *runner) runCharmHookOnLocal(hookName string, env []string, charmLo
...
 		}
 	}
-	return errors.Trace(exitErr)
+
+	return hookHandlerType, errors.Trace(exitErr)
+}
+
+// getHook checks to see if the dispatch hook exists, if not, check for the
+// given hookName.
+func (runner *runner) getHook(hookName, charmDir, charmLocation string) (HookHandlerType, string, error) {
```
Comment (achilleasa) [inline worker/uniter/runner/runner.go]: We should also rename this to something more obvious (e.g. lookup [or discover] hook handler script)

File: worker/uniter/runner/runner.go [hunk 3]
```diff
@@ -549,10 +596,27 @@ func (runner *runner) runCharmHookOnLocal(hookName string, env []string, charmLo
...
+	hook, err := searchHook(charmDir, hookDispatcherScript)
```
Comment (achilleasa) [inline worker/uniter/runner/runner.go]: LIkewise, `searchHook` is a bit misleading as to its purpose.

File: worker/uniter/runner/runner.go [hunk 4]
```diff
@@ -391,14 +429,20 @@ func (b *bufferAdaptor) Messagef(isPrefix bool, message string, args ...interfac
...
-func (runner *runner) runCharmHookOnRemote(hookName string, env []string, charmLocation string) error {
+func (runner *runner) runCharmHookOnRemote(hookName string, env []string, charmLocation string) (HookHandlerType, error) {
 	charmDir := runner.paths.GetCharmDir()
 	hook := filepath.Join(charmDir, filepath.Join(charmLocation, hookName))
 
+	// TODO(achilleasa): we need to update the remote executor to recognize
+	// the presence of the hook dispatcher script and report that in its
+	// response payload. As this requires several changes to other packages,
+	// we will assume for now that we are always using the legacy hook scripts.
+	hookHandlerType := LegacyHookHander
```
Comment (achilleasa) [inline worker/uniter/runner/runner.go]: We probably need to implement some mechanism for detecting the hook handler type that was used by the remote before this PR can land.

Comment (hmlanigan) [reply worker/uniter/runner/runner.go]: This will be covered in a follow up PR to address k8s actions using dispatch.

**Comment by hmlanigan** [comment]
forced push to rebase to develop with jujuc changes.

File: worker/uniter/runner/debug/server.go [hunk 5]
```diff
@@ -69,12 +74,19 @@ func (s *ServerSession) RunHook(hookName, charmDir string, env []string) error {
...
 		// exited uncleanly.
 		waitClientExit(s)
-		proc.Kill()
+		_ = proc.Kill()
 	}(cmd.Process)
 	return cmd.Wait()
 }
 
-func (s *ServerSession) writeDebugFiles(debugDir string) error {
+func findRunHookString(hookName, hookRunner string) string {
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply worker/uniter/runner/debug/server.go]: renamed to buildRunHookCmd

File: worker/uniter/runner/runner.go [hunk 6]
```diff
@@ -42,14 +42,47 @@ const (
 	runOnRemote
 )
 
+// HookHandlerType is used to indicate the type of script used for handling a
+// particular hook type.
+type HookHandlerType uint8
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply worker/uniter/runner/runner.go]: changing to a string.

File: worker/uniter/runner/runner.go [hunk 7]
```diff
@@ -42,14 +42,47 @@ const (
...
+
+// String implements fmt.Stringer for HookHandlerType.
+func (t HookHandlerType) String() string {
+	switch t {
+	case LegacyHookHander:
+		return "legacy, bespoke hook script"
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply worker/uniter/runner/runner.go]: I changed to ExplicitHookHandler.

File: worker/uniter/runner/runner.go (hunk 2 above)
Comment (hmlanigan) [reply worker/uniter/runner/runner.go]: renamed to discoverHookHandler

File: worker/uniter/runner/runner.go (hunk 3 above)
Comment (hmlanigan) [reply worker/uniter/runner/runner.go]: renamed to discoverHookScript.

**Comment by hmlanigan** [comment]
!!build!!

**Comment by hmlanigan** [comment]
$$merge$$

**Pull request merged**